### PR TITLE
Simplify plug-in API

### PR DESCRIPTION
**Pull request by bernardnormier**
This PR simplifies the plug-in API in C#:
 - it removes AddPlugin
 - it replaces all the GetXxx methods by a simple get-only Plugins property

### CODE REVIEW

File: csharp/src/Ice/Communicator.cs [hunk 1]
```diff
@@ -1297,6 +1306,39 @@ internal void UpdateConnectionObservers()
...
+            {
+                foreach ((string name, IPlugin plugin) in _plugins)
+                {
+                    await plugin.ActivateAsync(cancel).ConfigureAwait(false);
+                    activatedPlugins.Add(plugin);
+                }
+            }
+            catch (Exception)
+            {
+                // Destroy the plug-ins that have been successfully activated, in the reverse order.
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs:1326]: I think we should add a TODO to just destroy the plug-ins during DisposeAsync instead.

File: csharp/src/Ice/Communicator.cs [hunk 2]
```diff
@@ -210,6 +215,8 @@ internal CancellationToken CancellationToken
 
         private static string[] _emptyArgs = Array.Empty<string>();
 
+        private static readonly Dictionary<string, IPluginFactory> _pluginFactories = new();
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs:218]: This is no longer required, it was just to hold the factories added using the old API

File: csharp/src/Ice/Communicator.cs [hunk 3]
```diff
@@ -1404,6 +1446,216 @@ private static void LoadReferencedAssemblies(Assembly a)
...
+                    properties.ParseArgs(ref cmdArgs, name);
+                    foreach (KeyValuePair<string, string> p in properties)
+                    {
+                        SetProperty(p.Key, p.Value);
+                    }
+                }
+                Debug.Assert(entryPoint != null);
+                // Always check the static plugin factory table first, it takes precedence over the entryPoint specified
+                // in the plugin property value.
+                if (!_pluginFactories.TryGetValue(name, out IPluginFactory? pluginFactory))
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs:1576]: We no longer need to check `_pluginFactories` see previous comment

File: csharp/src/Ice/Communicator.cs [hunk 4]
```diff
@@ -1404,6 +1446,216 @@ private static void LoadReferencedAssemblies(Assembly a)
...
+                        // First try to load the assembly using Assembly.Load, which will succeed if a fully-qualified
+                        // name is provided or if a partial name has been qualified in configuration. If that fails, try
+                        // Assembly.LoadFrom(), which will succeed if a file name is configured or a partial name is
+                        // configured and DEVPATH is used.
+                        //
+                        // We catch System.Exception as this can fail with System.ArgumentNullException or
+                        // System.IO.IOException depending of the .NET framework and platform.
+                        try
+                        {
+                            pluginAssembly = System.Reflection.Assembly.Load(assemblyName);
```
Comment (pepone) [inline csharp/src/Ice/Communicator.cs:1611]: `System.Reflection.` is no needed in this file, we have a using directive for it.

**Review by pepone — APPROVED** [review]
Looks good, I wonder if would be better if LoadPlugin, returns an IPlugin and LoadPlugins an IReadOnlyList<IPlugin>, and make Plugins a read-ony property, then the constructor can call `LoadPlugins` to assign it

```
Plugins = LoadPlugins(ref args);

...

IReadOnlyList<IPlugin> LoadPlugins(ref string[] cmdArgs){...}

IPlugin LoadPlugin(string name, string pluginSpec, ref string[] cmdArgs) {...}
```

File: csharp/src/Ice/Communicator.cs (hunk 1 above)
Comment (pepone) [reply csharp/src/Ice/Communicator.cs:1326]: actually, we are calling twice DisposeAsync in each plug-in one here, and again in Communicator.DisposeAsync

**Comment by bernardnormier** [comment]
> Looks good, I wonder if would be better if LoadPlugin, returns an IPlugin and LoadPlugins an IReadOnlyList, and make Plugins a read-ony property, then the constructor can call `LoadPlugins` to assign it

This means a plug-in would not be able to check which plug-ins are already loaded when it's loaded. That's why I kept as-is.

File: csharp/src/Ice/Communicator.cs (hunk 4 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator.cs:1611]: Fixed.

File: csharp/src/Ice/Communicator.cs (hunk 3 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator.cs:1576]: Fixed.

File: csharp/src/Ice/Communicator.cs (hunk 2 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator.cs:218]: Fixed.

File: csharp/src/Ice/Communicator.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/Communicator.cs:1326]: I just fixed it and removed this method.

**Pull request merged**